### PR TITLE
HSB-445 feature: storing user last login timestamp

### PR DESCRIPTION
**Pull request by mirarifhasan**
Closes HSB-444

### Description

This PR introduces the ability to store the user's last login timestamp in `hoppscotch-backend` package (in the `user` table). `lastLoggedOn` is the property to access the data. This value can be nullable for previous data.

### Checks

- [x] My pull request adheres to the code style of this project
- [ ] My code requires changes to the documentation
- [ ] I have updated the documentation as required
- [x] All the tests have passed

### Additional Information

Nil

### CODE REVIEW

File: packages/hoppscotch-backend/src/user/user.service.ts [hunk 1]
```diff
@@ -316,6 +320,17 @@ export class UserService {
     }
   }
 
+  /**
+   * Update user's lastLoggedOn timestamp
```
Comment (balub) [inline packages/hoppscotch-backend/src/user/user.service.ts:324]: Can you rewrite this function to handle errors too, it currently is not a unit function.

File: packages/hoppscotch-backend/src/user/user.service.ts [hunk 2]
```diff
@@ -141,6 +142,7 @@ export class UserService {
     const createdUser = await this.prisma.user.create({
       data: {
         email: email,
+        lastLoggedOn: new Date(),
```
Comment (balub) [inline packages/hoppscotch-backend/src/user/user.service.ts:145]: I dont think we should update the time here, the user is techincally not logged-in at this point.

File: packages/hoppscotch-backend/src/user/user.service.ts [hunk 3]
```diff
@@ -114,14 +114,15 @@ export class UserService {
...
-  async UpdateUserRefreshToken(refreshTokenHash: string, userUid: string) {
+  async updateUserRefreshToken(refreshTokenHash: string, userUid: string) {
     try {
       const user = await this.prisma.user.update({
         where: {
           uid: userUid,
         },
         data: {
           refreshToken: refreshTokenHash,
+          lastLoggedOn: new Date(),
```
Comment (balub) [inline packages/hoppscotch-backend/src/user/user.service.ts:125]: Here too, the user isnt techincally logged-in when the refresh tokens are updated.

File: packages/hoppscotch-backend/src/auth/auth.service.spec.ts [hunk 4]
```diff
@@ -172,7 +173,7 @@ describe('verifyMagicLinkTokens', () => {
     // generateAuthTokens
     mockJWT.sign.mockReturnValue(user.refreshToken);
     // UpdateUserRefreshToken
-    mockUser.UpdateUserRefreshToken.mockResolvedValueOnce(E.right(user));
+    mockUser.updateUserRefreshToken.mockResolvedValueOnce(E.right(user));
```
Comment (balub) [inline packages/hoppscotch-backend/src/auth/auth.service.spec.ts:176]:
```suggestion
    mockUser.updateUserLastLoggedOn.mockResolvedValueOnce(E.right(user));
```

File: packages/hoppscotch-backend/src/auth/auth.service.spec.ts [hunk 5]
```diff
@@ -197,7 +198,7 @@ describe('verifyMagicLinkTokens', () => {
...
```
Comment (balub) [inline packages/hoppscotch-backend/src/auth/auth.service.spec.ts:201]:
```suggestion
    mockUser.updateUserLastLoggedOn.mockResolvedValueOnce(E.right(user));
```

File: packages/hoppscotch-backend/src/auth/auth.service.spec.ts [hunk 6]
```diff
@@ -239,7 +240,7 @@ describe('verifyMagicLinkTokens', () => {
     // generateAuthTokens
     mockJWT.sign.mockReturnValue(user.refreshToken);
     // UpdateUserRefreshToken
-    mockUser.UpdateUserRefreshToken.mockResolvedValueOnce(
+    mockUser.updateUserRefreshToken.mockResolvedValueOnce(
```
Comment (balub) [inline packages/hoppscotch-backend/src/auth/auth.service.spec.ts:243]:
```suggestion
    mockUser.updateUserLastLoggedOn.mockResolvedValueOnce(
```

File: packages/hoppscotch-backend/src/auth/auth.service.spec.ts [hunk 7]
```diff
@@ -264,7 +265,7 @@ describe('verifyMagicLinkTokens', () => {
     // generateAuthTokens
     mockJWT.sign.mockReturnValue(user.refreshToken);
     // UpdateUserRefreshToken
-    mockUser.UpdateUserRefreshToken.mockResolvedValueOnce(E.right(user));
+    mockUser.updateUserRefreshToken.mockResolvedValueOnce(E.right(user));
```
Comment (balub) [inline packages/hoppscotch-backend/src/auth/auth.service.spec.ts:268]:
```suggestion
    mockUser.updateUserLastLoggedOn.mockResolvedValueOnce(E.right(user));
```

File: packages/hoppscotch-backend/src/auth/auth.service.spec.ts [hunk 8]
```diff
@@ -197,7 +198,7 @@ describe('verifyMagicLinkTokens', () => {
...
     // deletePasswordlessVerificationToken
     mockPrisma.verificationToken.delete.mockResolvedValueOnce(passwordlessData);
```
Comment (balub) [inline packages/hoppscotch-backend/src/auth/auth.service.spec.ts:204]: Missing `updateUserLastLoggedOn` and in other related testcases.

File: packages/hoppscotch-backend/src/user/user.service.ts (hunk 3 above)
Comment (balub) [reply packages/hoppscotch-backend/src/user/user.service.ts:125]: Also are we sure the date is in the correct format expected by the database.

Comment (mirarifhasan) [reply packages/hoppscotch-backend/src/user/user.service.ts:125]: The date format is the same as `createdOn` perspective of DB. So, it's good.

**Review by balub — CHANGES_REQUESTED** [review]
We are not calling `updateUserLastLoggedOn` in the strategies for the SSO providers. It currently does not update the last logged in when a user logs in via any SSO provider.

**Review by balub — APPROVED** [review]
LGTM 🚀

**Pull request merged**